Telemetry Documentation (#16917)
mlflow/mlflow: docs/docs/community/usage-tracking.mdx

    </tr>
    <tr>
      <td>Tracking URI Scheme</td>
      <td>The scheme of the current tracking URI</td>
      <td>file | sqlite | mysql | postgresql | mssql | https | http | custom_scheme | None</td>
      <td>To determine which tracking backends are most commonly used and optimize backend support</td>
    </tr>
    <tr>
      <td>Event Name</td>
      <td>The tracked event name (see [below table](#tracked-events) for what events are tracked)</td>
      <td>create_experiment</td>
      <td>To measure feature usage and improvements</td>
    </tr>
    <tr>
      <td>Event Status</td>
      <td>Whether the event succeeds or not</td>
      <td>success | failure | unknown</td>
      <td>To identify common failure points and improve reliability and error handling</td>
    </tr>
    <tr>
      <td>Duration</td>
      <td>The time the event call takes, in milliseconds</td>
      <td>1000</td>
      <td>To monitor performance trends and detect regressions in response time</td>
    </tr>
    <tr>
      <td>Parameters (boolean or enumerated values)</td>
      <td>See [below table](#tracked-events) for collected parameters for each event</td>

>       <td>-</td>

Reviewer: can we put an example here?

      <td>To better understand the usage pattern for each event</td>
    </tr>
  </tbody>
</Table>

#### Tracked Events

**No details about the specific model, code, or weights are collected.** Only the parameters listed under the `Tracked Parameters` column are recorded alongside the event;
For events with None in the `Tracked Parameters` column, only the event name is recorded.

<Table>
  <thead>
    <tr>
      <th>Event Name</th>

>       <th>Tracked Parameters</th>

Reviewer: > only two events have examples

I think it's still worth it.

    </tr>
  </thead>
  <tbody>
    <tr>

>       <td>create_experiment</td>

Reviewer: can we add the `import_mlflow` event?

      <td>None</td>
    </tr>
    <tr>
      <td>create_run</td>

>       <td>Whether the packages in [PACKAGES_TO_CHECK_IMPORT](https://github.com/mlflow/mlflow/blob/c71fd0d677c1806ba2d5928398435c4de2c25c0e/mlflow/telemetry/constant.py#L19) is imported or not</td>

Reviewer: ```suggestion
      <td>Whether the packages in [PACKAGES_TO_CHECK_IMPORT](https://github.com/mlflow/mlflow/blob/c71fd0d677c1806ba2d5928398435c4de2c25c0e/mlflow/telemetry/constant.py#L19) are imported or not</td>
```

- Measure the impact of changes and ensure they improve the experience for the broader community

If telemetry were opt-in, only a small, self-selected subset of users would be represented, leading to biased insights and potentially misaligned priorities.
We are committed to transparency and user choice. Telemetry is clearly documented, anonymized, and can be easily disabled at any time through configuration.
This approach helps us make MLflow better for everyone, while giving you full control. Check [`what we are doing with this data`](#what-are-we-doing-with-this-data) section for more information.

## How to opt-out?

MLflow supports opt-out telemetry through either of the following environment variables:

- **MLFLOW_DISABLE_TELEMETRY=true**
- **DO_NOT_TRACK=true**

Setting either of these will **immediately disable telemetry**, no need to re-import MLflow or restart your session.

:::note
MLflow automatically disables telemetry in the following CI environments.
If you’d like support for additional CI environments, please [open an issue on our GitHub repository](https://github.com/mlflow/mlflow/issues).

- CI
- Github Actions
- Circle CI
- GitLab CI
- Jenkins
- Travis

> - Azure Pipeline

Reviewer: ```suggestion
- Azure Pipelines
```

- **User Control**: Users can opt out of telemetry at any time by setting the environment variable MLFLOW_DISABLE_TELEMETRY=true or DO_NOT_TRACK=true. MLflow respects these settings immediately without requiring a restart.
- **Transparency**: Telemetry endpoints and behavior are documented publicly, and MLflow users can inspect or block the relevant network calls.

For further inquiries or data protection questions, users can file an issue on the [MLflow GitHub repository](https://github.com/mlflow/mlflow/issues).

## What data is collected?

MLflow collects only non-sensitive, anonymized data to help us better understand usage patterns.
The below section outlines the data currently collected in this version of MLflow. You can view the exact data collected [in the source code](https://github.com/mlflow/mlflow/blob/c71fd0d677c1806ba2d5928398435c4de2c25c0e/mlflow/telemetry/schemas.py).

### Data Explanation

<Table>
  <thead>
    <tr>
      <th>Data Element</th>
      <th>Explanation</th>
      <th>Example</th>
      <th>Why we track this</th>
    </tr>
  </thead>
  <tbody>
    <tr>

>       <td>Unique session ID</td>

Reviewer: Is `timestamp_ns` missing?

Author: Thanks for the catch!

If telemetry were opt-in, only a small, self-selected subset of users would be represented, leading to biased insights and potentially misaligned priorities.
We are committed to transparency and user choice. Telemetry is clearly documented, anonymized, and can be easily disabled at any time through configuration.
This approach helps us make MLflow better for everyone, while giving you full control. Check [`what we are doing with this data`](#what-are-we-doing-with-this-data) section for more information.

## How to opt-out?

MLflow supports opt-out telemetry through either of the following environment variables:

- **MLFLOW_DISABLE_TELEMETRY=true**
- **DO_NOT_TRACK=true**

Setting either of these will **immediately disable telemetry**, no need to re-import MLflow or restart your session.

:::note
MLflow automatically disables telemetry in the following CI environments.
If you’d like support for additional CI environments, please [open an issue on our GitHub repository](https://github.com/mlflow/mlflow/issues).

- CI
- Github Actions
- CircleCI
- GitLab CI
- Jenkins
- Travis
- Azure Pipeline
- BitBucket
- CodeBuild
- BuildKite
  :::

> 

Reviewer: Can we provide a code example?

Reviewer: Looks like the new GitHub UI is messed up. Let me repost the comment.

### Scope of the setting

- The environment variable only takes effect in processes where it is explicitly set or inherited.

> - If you spawn subprocesses (e.g., via subprocess.run) from a clean environment, those subprocesses may not inherit your shell’s environment, and telemetry could remain enabled.

Reviewer: https://github.com/mlflow/mlflow/pull/16917/files#r2238307595

Author: For showing when the environment variables are not inherited?

Reviewer: Yep

Recommendations to ensure telemetry is consistently disabled across all environments:

- Add the variable to your shell startup file (~/.bashrc, ~/.zshrc, etc.): `export MLFLOW_DISABLE_TELEMETRY=true`

> - If you’re using subprocesses or isolated environments, use a dotenv manager or explicitly pass the variable when launching.

Author: That is not as accurate as get_telemetry_client, it's possible that it's not disabled in the environment but our config disables the telemetry. Also I don't want to expose `is_telemetry_disabled` as a public facing API

> - If you’re using subprocesses or isolated environments, use a dotenv manager or explicitly pass the variable when launching.

Reviewer: I see, didn't know that.

### How to validate telemetry is disabled?

Use `mlflow.telemetry.get_telemetry_client()` to retrieve the current telemetry client. If it returns None, telemetry is disabled for the current session.

### How to opt-out for your organization?

> MLflow fetches telemetry configuration from `https://config.mlflow-telemetry.io`.
> When it fails to fetch the configuration, MLflow will not track any telemetry for the session.
> With this, you can block requests to this endpoint to prevent telemetry from being enabled.

Reviewer: ```suggestion
Organizations can disable telemetry by blocking network access to `https://config.mlflow-telemetry.io`. When this endpoint is unreachable, MLflow automatically disables telemetry.
```

- **DO_NOT_TRACK=true**

Setting either of these will **immediately disable telemetry**, no need to re-import MLflow or restart your session.

:::note
MLflow automatically disables telemetry in the following CI environments.
If you’d like support for additional CI environments, please [open an issue on our GitHub repository](https://github.com/mlflow/mlflow/issues).

- CI
- Github Actions
- CircleCI
- GitLab CI/CD
- Jenkins Pipeline
- Travis CI
- Azure Pipelines
- BitBucket
- AWS CodeBuild
- BuildKite
  :::

### Scope of the setting

- The environment variable only takes effect in processes where it is explicitly set or inherited.
- If you spawn subprocesses from a clean environment, those subprocesses may not inherit your shell’s environment, and telemetry could remain enabled. e.g. `subprocess.run([...], env={})`

Recommendations to ensure telemetry is consistently disabled across all environments:

- Add the variable to your shell startup file (~/.bashrc, ~/.zshrc, etc.): `export MLFLOW_DISABLE_TELEMETRY=true`
- If you’re using subprocesses or isolated environments, use a dotenv manager or explicitly pass the variable when launching.

### How to validate telemetry is disabled?

> Use `mlflow.telemetry.get_telemetry_client()` to retrieve the current telemetry client. If it returns None, telemetry is disabled for the current session.

Reviewer: ````suggestion
Use the following code to validate telemetry is disabled.

```python
from mlflow.telemetry import get_telemetry_client

assert get_telemetry_client() is None, "Telemetry is enabled"
```
````

can we provide an example?
